File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    answer = []
    name = {}
    temp = []
    for r in  record:
        lst = r.split()
        if len(lst) == 2:
            temp.append(lst[1]+'!'+"님이 나갔습니다.")
        else:
            name[lst[1]] = lst[2]
            if lst[0] == "Enter":
                temp.append(lst[1]+'!'+"님이 들어왔습니다.")
    for t in temp:
        lst2 = t.split('!')
        key = lst2[0]
        lst2[0] = name[key]
        answer.append(''.join(lst2))
            
    return answer

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;
class Solution {
    public String[] solution(String[] record) {
        List<String> answer = new ArrayList<>();
        HashMap<String, String> map = new HashMap<>();
        
        for (String r : record) {
            String[] message = r.split(" ");
            if (!r.startsWith("Leave")) {
                map.put(message[1], message[2]);
            }
        }
        
        for (String r : record) {
            String[] message = r.split(" ");
            if (r.startsWith("Enter")) {
                answer.add(map.get(message[1])+"님이 들어왔습니다.");
            } else if (r.startsWith("Leave")) {
                answer.add(map.get(message[1])+"님이 나갔습니다.");
            }
        }
        return answer.toArray(new String[answer.size()]);
    }
}